Enhancement/29949 integrate users and groups into inventory harvester

MiguelazoDS opened the pull request:
|Related issue|
|---|
|#799|

## Description

Implementation of global queries for two new providers: 
- Users
- Groups
- Uts
- QA tests

GabrielEValenzuela commented on src/wazuh_modules/inventory_harvester/qa/test_data/inventory/users/000_insert_rsync/inputs/1_rsync.json:
```diff
@@ -0,0 +1,48 @@
+            "user_uuid": null
+        }
+    }
+}
```
We have a missing endline

GabrielEValenzuela commented on src/wazuh_modules/inventory_harvester/qa/test_data/inventory/users/002_insert_delta/result.json:
```diff
@@ -0,0 +1,74 @@
+            }
+        ]
+    }
+]
```
Ditto

GabrielEValenzuela reviewed (changes requested):
# Testing 🔴 

I follow the same procedure that described [here](https://github.com/wazuh/wazuh/pull/30404#pullrequestreview-2955250689) but now analyzing the indexer.

> [!CAUTION]
> I used a `4.12.0` and then overwrite server with this version, please, disable VD modules to avoid errors due incompatibility of the content 

### Index creation 🟢 
ligue
```console
vboxuser@vboxuser:~/wazuh$ curl -k -u admin:Vd4GJeEYLAHdowWAAdXv5?*5BuF.YIXL https://127.0.0.1:9200/_cat/indices/wazuh-*?v
health status index                                      uuid                   pri rep docs.count docs.deleted store.size pri.store.size
green  open   wazuh-states-fim-files-vboxuser            zKjobQ11Sw66p5gxsbgduw   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-protocols-vboxuser  15ED51ueQmy0gCcGVksWNQ   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-ports-vboxuser      VXsU2_R_Q4KUQHjjLtXcGw   1   0          0            0       208b           208b
green  open   wazuh-states-fim-registries-vboxuser       3C_qTwo7S6CyE4tk_48nGg   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-hotfixes-vboxuser   yyz0JAlEQqKq-iP9obmpQw   1   0          0            0       208b           208b
green  open   wazuh-states-vulnerabilities-vboxuser      iTsJv4fwSTSr1U9xqfRCwA   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-users-vboxuser      42eUul1uRZiFaMN4c90CXQ   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-hardware-vboxuser   pjkshLA0Q2-anHE5iWNQqQ   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-groups-vboxuser     YsscwxPTTR23yTkioxX6tg   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-system-vboxuser     NSxJ8d9ORXK-iodLpGDqHw   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-packages-vboxuser   VwJRhomBQDOBGKuynUdGPg   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-processes-vboxuser  VjfAfw9nSeKdJXj8l3ifNg   1   0          0            0       208b           208b
green  open   wazuh-states-inventory-interfaces-vboxuser sV6YrqGqSpqWOCbA-kU7_w   1   0          0            0       208b           208b
green  open   wazuh-alerts-4.x-2025.06.26                FODfbqCgRKCRW8TDp5AtxQ   3   0        301            0        1mb            1mb
green  open   wazuh-states-inventory-networks-vboxuser   FeP5jM6iRDmYvAFX2vx9pg   1   0          0            0       208b           208b
green  open   wazuh-monitoring-2025.26w                  xakr73AlR8yiH-qGaC99pQ   1   0          0            0       208b           208b
```

#### Index insertion 🔴 

I attach the JSON files, the user `batman` and group `justice_ligue` are in the DB but no on index

#### Delete operation 🟢 

Username is deleted from database

> [!NOTE]
> For Linux I used delta messages

--------------------

For Windows scenario (RSync) I found the same behavior. I attach the `ossec.log`

----

[data.zip](https://github.com/user-attachments/files/20929809/data.zip)

MiguelazoDS commented on src/shared_modules/utils/timeHelper.h:
```diff
@@ -200,7 +201,7 @@ namespace Utils
     }
 
-    static std::string rawTimestampToISO8601(const uint32_t timestamp)
+    static std::string rawTimestampToISO8601(const double timestamp)
```
I added more tests, do you think still there's a change of failure? I extended the precision of the timestamp because of new providers.

GabrielEValenzuela replied:
We solve this a call 🚀

MiguelazoDS commented:
# Linux 

## Logs 
[ossec.log](https://github.com/user-attachments/files/21005920/ossec.log)

No error or logs to be worried about 
```console
 grep -E "ERR|WARN|CRIT" ossec.log | grep -v "wm_sca"
2025/07/01 15:08:03 indexer-connector[137594] indexerConnector.cpp:161 at initConfiguration(): WARNING: No username and password found in the keystore, using default values.
```

No indexing errors 
```console
grep -E "errors\":true" ossec.log | wc -l
0
grep -E "errors\":false" ossec.log | wc -l
54
```

## Initial scan 

Dbs

```console
root@jammy:/home/vagrant/wazuh# sqlite3 /var/ossec/queue/db/001.db 'select * from sys_users' | wc -l
16
root@jammy:/home/vagrant/wazuh# sqlite3 /var/ossec/queue/db/001.db 'select * from sys_groups' | wc -l
31
```

[groups_db.log](https://github.com/user-attachments/files/21005907/groups_db.log)
[users_db.log](https://github.com/user-attachments/files/21005908/users_db.log)

Inventory

![image](https://github.com/user-attachments/assets/dd1dece1-6870-4691-991c-0d96b723f44a)

[groups1.json](https://github.com/user-attachments/files/21005949/groups1.json)
[users1.json](https://github.com/user-attachments/files/21005950/users1.json)

## Creating docker group and assign it to new user wazuh_user1 

New user
```console
                                     scan_id = 0
                                   scan_time = 2025/07/01 18:36:15
                                   user_name = wazuh_user1
                              user_full_name =
                                   user_home = /home/wazuh_user1
                                     user_id = 1000
                             user_uid_signed = 1000
                                   user_uuid =
                                 user_groups = root:docker
                               user_group_id = 0
                        user_group_id_signed = 0
                                user_created =
                                  user_roles =
                                  user_shell = /bin/bash
                                   user_type =
                              user_is_hidden = 0
                              user_is_remote = 1
                             user_last_login =
                      user_auth_failed_count = 0
                  user_auth_failed_timestamp =
                 user_password_last_set_time =
               user_password_expiration_date =
                user_password_hash_algorithm =
                 user_password_inactive_days =
                   user_password_last_change = 20270
      user_password_max_days_between_changes = 99999
      user_password_min_days_between_changes = 0
                        user_password_status = not_set
user_password_warning_days_before_expiration = 7
                                 process_pid = 0
                                     host_ip =
                                login_status = 0
                                   login_tty =
                                  login_type =
                                    checksum = 40a110feffbabb9aedbaf7af7b6a62d9e2d33661
```

```json
{
        "_index": "wazuh-states-inventory-users-jammy",
        "_id": "001_wazuh_user1",
        "_score": 1,
        "_source": {
          "agent": {
            "id": "001",
            "name": "c9cc11ad8298",
            "version": "v4.14.0"
          },
          "login": {
            "status": false
          },
          "process": {
            "pid": 0
          },
          "user": {
            "auth_failures": {
              "count": 0
            },
            "group": {
              "id": 0,
              "id_signed": 0
            },
            "groups": [
              "root",
              "docker"
            ],
            "home": "/home/wazuh_user1",
            "id": "1000",
            "is_hidden": false,
            "is_remote": true,
            "name": "wazuh_user1",
            "password": {
              "last_change": 20270,
              "max_days_between_changes": 99999,
              "min_days_between_changes": 0,
              "status": "not_set",
              "warning_days_before_expiration": 7
            },
            "shell": "/bin/bash",
            "uid_signed": 1000
          },
          "wazuh": {
            "cluster": {
              "name": "jammy"
            },
            "schema": {
              "version": "1.0"
            }
          }
        }
      }

```

New group
```console
          scan_id = 0
        scan_time = 2025/07/01 18:36:15
         group_id = 1000
       group_name = docker
group_description =
  group_id_signed = 1000
       group_uuid =
  group_is_hidden = 0
      group_users = wazuh_user1
         checksum = 97929248ed2c405e1252937858dfbad6f739a693
```

```json
      {
        "_index": "wazuh-states-inventory-groups-jammy",
        "_id": "001_docker",
        "_score": 1,
        "_source": {
          "agent": {
            "id": "001",
            "name": "c9cc11ad8298",
            "version": "v4.14.0"
          },
          "group": {
            "id": 1000,
            "id_signed": 1000,
            "is_hidden": false,
            "name": "docker",
            "users": [
              "wazuh_user1"
            ]
          },
          "wazuh": {
            "cluster": {
              "name": "jammy"
            },
            "schema": {
              "version": "1.0"
            }
          }
        }
      }
``` 

## Deleting docker and wazuh_user1 

User and group successfully removed from database and indexer 

```console
2025/07/01 15:30:10 wazuh-remoted: INFO: AGENT EVENT: {"agent_info":{"agent_id":"001","agent_ip":"any","agent_name":"c9cc11ad8298","agent_version":"v4.14.0"},"data_type":"dbsync_users","data":{"checksum":"40a110feffbabb9aedbaf7af7b6a62d9e2d33661","host_ip":" ","login_status":0,"login_tty":" ","login_type":" ","process_pid":0,"scan_time":"2025/07/01 18:30:07","user_auth_failed_count":0,"user_auth_failed_timestamp":0,"user_created":0,"user_full_name":null,"user_group_id":0,"user_group_id_signed":0,"user_groups":"root:docker","user_home":"/home/wazuh_user1","user_id":1000,"user_is_hidden":0,"user_is_remote":1,"user_last_login":0,"user_name":"wazuh_user1","user_password_expiration_date":-1,"user_password_hash_algorithm":null,"user_password_inactive_days":-1,"user_password_last_change":20270,"user_password_last_set_time":0,"user_password_max_days_between_changes":99999,"user_password_min_days_between_changes":0,"user_password_status":"not_set","user_password_warning_days_before_expiration":7,"user_roles":" ","user_shell":"/bin/bash","user_type":" ","user_uid_signed":1000,"user_uuid":" "},"operation":"DELETED"}
```

```console
2025/07/01 15:30:08 wazuh-remoted: INFO: AGENT EVENT: {"agent_info":{"agent_id":"001","agent_ip":"any","agent_name":"c9cc11ad8298","agent_version":"v4.14.0"},"data_type":"dbsync_groups","data":{"checksum":"97929248ed2c405e1252937858dfbad6f739a693","group_description":" ","group_id":1000,"group_id_signed":1000,"group_is_hidden":0,"group_name":"docker","group_users":"wazuh_user1","group_uuid":" ","scan_time":"2025/07/01 18:30:07"},"operation":"DELETED"}
```

```console
2025/07/01 15:30:10 indexer-connector[137594] indexerConnector.cpp:615 at operator()(): DEBUG: Added document for deletion with id: 001_wazuh_user1.
2025/07/01 15:30:10 indexer-connector[137594] indexerConnector.cpp:615 at operator()(): DEBUG: Added document for deletion with id: 001_docker.
```

MiguelazoDS commented:
# Coverity 

```console
Done building server

Attempting to detect unconfigured compilers in build
|0----------25-----------50----------75---------100|
****************************************************
Emitted 552 C/C++ compilation units (100%) successfully

552 C/C++ compilation units (100%) are ready for analysis
The cov-build utility completed successfully.
Version: 4.14.0-alpha0
Branch: enhancement/29949-integrate-users-and-groups-into-inventory-harvester (8f078415462440102de21a1b37d8e45c461c4d97)
Project: wazuh%2Fossec-wazuh (lsjToeeRLIhcka9PXdqhvQ)
Target: server

Do you want to upload? [Y/n]: Y
Build successfully submitted.
 Please ensure that 'cov-build version' matches one of these: 2024.6.1 | 2023.12.2
 ```

MiguelazoDS commented:
# MacOS 

## Log 

## Initial scan 

## Creating dummy_group group and assign it to new user dummy_user 

## Deleting dummy_group and dummy_user

JcabreraC reviewed (changes requested):
Update `last_change` to store full epoch timestamp in seconds instead of days (e.g., use 1676592000 instead of 19405) to avoid manual conversion and improve date handling.

MiguelazoDS commented:
# MacOS 

After changes from agent and indexer that motivated the refactor of some fields, I'll perform a new exploratory testing to validate the changes.

GabrielEValenzuela commented:
# Update

Testing is passing, we need to update the mapping of indexer to finish it

GabrielEValenzuela commented:
# Update

I've rebased this branch along with the agent branch and executed the test scenario, making sure to remove all agent data beforehand by deleting the agents.

Below are the JSON documents from both Wazuh DB and the indexer.

[data.zip](https://github.com/user-attachments/files/21064704/data.zip)


My approval stands. ✅

GabrielEValenzuela commented:
# Testing 🟢 

Information is indexed correctly

![image](https://github.com/user-attachments/assets/b32affd7-8772-4bc3-8623-968d33d1d684)


![image](https://github.com/user-attachments/assets/fce40d10-0c13-4897-84c8-05edd4f572f3)
